### DISCUSSION TITLE
accessing time/date from BACnet client

### DISCUSSION THREAD

**Discussion by timcurry1** [JoelBender/bacpypes#462] [General]
Does BACpypes support requesting the time and date from the client? I have issues with my RTC on our server and I want to get the time and date from another source like the client. If so, where can I find examples.

**Comment by JoelBender** [comment]
The local date and time are properties of the device object.  The **ReadProperty.py** sample application will present a ">" prompt, enter something like `read 12.23.34.45 device:999 localDate` and `localTime`.

The **ReadPropertyMultiple.py** application allows you to create one request like `read 12.23.34.45 device:999 localDate localTime` but the device will respond with an error if it doesn't support that service.

**Reply by timcurry1** [reply]
Joel,

I am running ReadProperty.py from my Linux computer. I am trying to request the time from YABI on my Windows Computer.
I entered  "read 192.168.86.33 device:599 localDate"
at the prompt but I get "no response."
Is the device 599 for the YABI on my Windows Computer? I don't know what YABI's device ID is. 
What ini file is ReadProperty.py using? 

Here is YABI's Log screen:
ConfirmedServiceRequest
Sending ErrorResponse ... 
Confirmed service not handled: SERVICE_CONFIRMED_READ_PROPERTY
ConfirmedServiceRequest
Sending ErrorResponse ... 
Confirmed service not handled: SERVICE_CONFIRMED_READ_PROPERTY
ConfirmedServiceRequest
Sending ErrorResponse ... 
Confirmed service not handled: SERVICE_CONFIRMED_READ_PROPERTY
ConfirmedServiceRequest
Sending ErrorResponse ... 
Confirmed service not handled: SERVICE_CONFIRMED_READ_PROPERTY

**Comment by JoelBender** [comment]
Send an "unconstrained Who-Is" -- which is to say one that does not have a low limit or a high limit -- directly to the device and it will respond with an I-Am that contains the device object identifier, for example, `whois 192.168.86.33`.  You can also send out an unconstrained Who-Is broadcast to the entire network like `whois *` and filter out the responses, but a unicast message is preferred.

**Comment by timcurry1** [comment]
You said " whois 192.168.86.33" or "whois *" to find my deviceId of YABE.
Do I enter this at the ReadProperty.py '>' prompt? I tried it and I got an
error.

On Tue, Jun 21, 2022 at 7:38 PM Joel Bender ***@***.***>
wrote:

> The local date and time are properties of the device object. The
> *ReadProperty.py* sample application will present a ">" prompt, enter
> something like read 12.23.34.45 device:999 localDate and localTime.
>
> The *ReadPropertyMultiple.py* application allows you to create one
> request like read 12.23.34.45 device:999 localDate localTime but the
> device will respond with an error if it doesn't support that service.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JoelBender/bacpypes/discussions/462#discussioncomment-2998035>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKJU6EJ7CMRO7KTR6NEKMTVQJOAZANCNFSM5ZOAVZIQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Tim Curry

**Reply by JoelBender** [reply]
Ah, wrong application, **WhoIsIAm.py** is the correct one.  The idea was to keep the suite of applications with each one focused on a particular service and if you wanted to merged them all together into a "master CLI" or something you could do that.  I've often considered making `python -m bacpypes` launch such a thing, or maybe that starts a relatively persistent service daemon and other applications use it, or maybe it runs a BACnet Web Services server... ah, so many options.

**Reply by timcurry1** [reply]
I like the idea of using a master thread to run all  of these commands. Maybe implement a state machine,  too.

I have YABE up. When I type 'whois 192.168.86.33', I get my ./WhoIsIAm.py prompt back and nothing is returned. I can ping 192.168.86.33. I see that YABE displays the phrase 'UnconfirmedServiceRequest' when I hit return. I think I see YABE's device ID = 400001 in an XML file. I should get that device ID. I also tried 'whois *' and I don't get anything back, either.

**Reply by JoelBender** [reply]
You are correct, when YABE receives a Who-Is request it should respond with an I-Am as a unicast or broadcast message.  Running Wireshark will tell you if it is, which probably means your configuration needs help or there is some other OS rule blocking it.  Please double check your configuration file that your address is like `12.23.34.45/24` (CIDR notation) to verify that it knows what network it's on, and you should also see through some OS tool that the application is listening on `12.23.34.255` for broadcast packets.
